GEOSldas_App/util/postproc/ObsFcstAna_stats/Main_example.py

```diff
@@ -0,0 +1,407 @@
+
+#  =========================================================================
+#  Postprocess raw ObsFcstAna output data into monthly sums for simpler and faster postprocessing;
+#  computes mean, variance from monthly sums that can be used to compute DA diagnostics directly
+
+if not os.path.isfile(stats_file):
+    # Initialize the postprocessing object
+    postproc = postproc_ObsFcstAna(exp_list, start_time, end_time, obs_from=obs_from)
+    # Step 1: Compute and save monthly sums 
+    postproc.save_monthly_sum(out_path_mo)
+    # Step 2: Compute statistics from monthly sums, option to save result to file
+    stats = postproc.calculate_stats_from_sums(mo_path=out_path_mo, write_to_nc=True, filename=stats_file)
+else:
+    print('reading stats nc4 file '+stats_file)
+    stats = {}
+    with Dataset(stats_file,'r') as nc:
+        for key, value in nc.variables.items():
+            stats[key] = value[:].filled(np.nan)
+# 
+#  ==========================================================================
+
+# Sample of final compuation of selected diagnostic metrics 
+ 
+Nmin = 20
+
+# Then computer metrics of O-F, O-A, etc. based on above computed
+N_data = stats['N_data']
+O_mean = stats['obs_mean']
+# mean(x-y) = E[x] - E[y]   
+OmF_mean = stats['obs_mean'] - stats['fcst_mean']
+OmA_mean = stats['obs_mean'] - stats['ana_mean']
+# var(x-y) = var(x) + var(y) - 2cov(x,y)
+# cov(x,y) = E[xy] - E[x]E[y]
+OmF_stdv  = np.sqrt(stats['obs_variance'] + stats['fcst_variance'] - \
+                    2 * (stats['oxf_mean'] - stats['obs_mean']*stats['fcst_mean']))
+                    
+OmA_stdv  = np.sqrt(stats['obs_variance'] + stats['ana_variance'] - \
+                    2 * (stats['oxa_mean'] - stats['obs_mean']*stats['ana_mean']))
+
+ # "fcstvar" is assumed constant here for convenience. Modify if necessary
+OmF_norm_mean = OmF_mean / np.sqrt(stats['obsvar_mean'] + stats['fcstvar_mean']) 
+OmF_norm_stdv = np.sqrt(OmF_stdv**2 / (stats['obsvar_mean'] + stats['fcstvar_mean']) )
+  
+# Mask out data points with insufficent observations using the Nmin threshold
+# Do NOT apply to N_data
+OmF_mean[     N_data < Nmin] = np.nan
+OmF_stdv[     N_data < Nmin] = np.nan
+OmF_norm_mean[N_data < Nmin] = np.nan
+OmF_norm_stdv[N_data < Nmin] = np.nan
+OmA_mean[     N_data < Nmin] = np.nan
+OmA_stdv[     N_data < Nmin] = np.nan
+
+# Combine metrics of individual species using weighted averaging
+OmF_mean      = np.nansum(OmF_mean     *N_data, axis=1)/np.nansum(N_data,axis=1)
+OmF_stdv      = np.nansum(OmF_stdv     *N_data, axis=1)/np.nansum(N_data,axis=1)
+OmF_norm_mean = np.nansum(OmF_norm_mean*N_data, axis=1)/np.nansum(N_data,axis=1)
+OmF_norm_stdv = np.nansum(OmF_norm_stdv*N_data, axis=1)/np.nansum(N_data,axis=1)
+OmA_mean      = np.nansum(OmA_mean     *N_data, axis=1)/np.nansum(N_data,axis=1)
+OmA_stdv      = np.nansum(OmA_stdv     *N_data, axis=1)/np.nansum(N_data,axis=1)
+Nobs_data     = np.nansum(              N_data, axis=1)
```

Comment:
Can the computation of Nobs_data be moved up, so it can be used for normalizing, rather than computing  "np.nansum(N_data,axis=1)" in each of the preceding 6 lines?